File: Company_wise_Mock/Microsoft/Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int cntA = 0, cntB = 0;
        ListNode *temp;
        
        // cnt nodes in A and B
        temp = headA;
        while(temp) {
            cntA++;
            temp = temp->next;
        }
        
        temp = headB;
        while(temp) {
            cntB++;
            temp = temp->next;
        }
        
        ListNode *A = headA, *B = headB;
        while(cntA >  cntB) {
            cntA--;
            A = A->next;
        }
        
        while(cntB > cntA) {
            cntB--;
            B = B->next;
        }
        
        ListNode *ans = NULL;
        while(A and B){
            if(A == B){
                ans = A;
                break;
            }
            
            A = A->next;
            B = B->next;
        }
        
        return ans;
    }
};

File: Daily Challenges/2021/March/Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        cnt1 = 0
        cnt2 = 0
        
        temp = headA
        while temp:
            cnt1 += 1
            temp = temp.next
        
        temp = headB
        while temp:
            cnt2 += 1
            temp = temp.next
        
        while cnt1 > cnt2:
            headA = headA.next
            cnt1 -= 1
        
        while cnt2 > cnt1:
            headB = headB.next
            cnt2 -= 1
        
        while headA and headB:
            if headA == headB:
                return headA
            
            headA = headA.next
            headB = headB.next
        
        return None
